libkatana_python_native/include/katana/python/NumbaSupport.h

```diff
@@ -0,0 +1,230 @@
+  }
+
+  constexpr static bool is_const = true;
+};
+
+template <typename Cls, typename Return, typename... Args>
+constexpr auto
+MemberFunctionBinderInferer(Return (Cls::*)(Args...)) {
+  return MemberFunctionBinderImpl<Cls, Return, Args...>{};
+}
+
+template <typename Cls, typename Return, typename... Args>
+constexpr auto
+MemberFunctionBinderInferer(Return (Cls::*)(Args...) const) {
+  return ConstMemberFunctionBinderImpl<Cls, Return, Args...>{};
+}
+
+template <auto member_func>
+using MemberFunctionBinder = decltype(MemberFunctionBinderInferer(member_func));
+
+}  // namespace detail
+
+/// Define a Python and numba method.
+/// This is used as follows:
+/// @code
+/// DefWithNumba&lt;&func>(cls, "func");
+/// @endcode
+/// If the method is overloaded, you must select a specific overload using
+/// pybind11::overload_cast().
+/// @code
+/// DefWithNumba&lt;pybind11:overload_cast&lt;const std::string&>(&func)>(cls, "func");
+/// @endcode
+///
+/// Numba interfacing does not (yet) support exposing overloading itself to
+/// Numba code.
+///
+/// \tparam func The function pointer @b value (not type).
+/// \param cls The python class object.
+/// \param name The name of the method in python.
+/// \param extra Extra arguments as you would pass to pybind11::class_::def().
+template <auto func, typename... Extra>
+constexpr void
+DefWithNumba(pybind11::module_& m, const char* name, const Extra&... extra) {
+  static_assert(
+      std::is_same_v<
+          typename detail::StaticFunctionBinder<func>::StaticFunctionPointer,
+          decltype(func)>);
+  detail::StaticFunctionBinder<func>::template def_method<func>(
+      m, name, extra...);
+}
+
+/// Define a Python and numba method.
+/// This is used as follows:
+/// @code
+/// DefWithNumba&lt;&Cls::method>(cls, "method");
+/// @endcode
+/// If the method is overloaded, you must select a specific overload using
+/// pybind11::overload_cast().
+/// @code
+/// DefWithNumba&lt;pybind11:overload_cast&lt;const std::string&>(&Cls::method)>(cls, "method");
+/// @endcode
+///
+/// Numba interfacing does not (yet) support exposing overloading itself to
+/// Numba code.
+///
+/// \tparam member_func The member function pointer @b value (not type).
+/// \param cls The Python class object.
+/// \param name The name of the method in python.
+/// \param extra Extra arguments as you would pass to pybind11::class_::def().
+template <auto member_func, typename Cls, typename... Extra>
+constexpr void
+DefWithNumba(
+    pybind11::class_<Cls>& cls, const char* name, const Extra&... extra) {
+  static_assert(std::is_same_v<
+                typename detail::MemberFunctionBinder<
+                    member_func>::MemberFunctionPointer,
+                decltype(member_func)>);
+  detail::MemberFunctionBinder<member_func>::template def_class_method<
+      member_func>(cls, name, extra...);
+}
+
+// TODO(amp): Implement missing overloads/variants for: lambdas (as methods and
+//  functions), static methods.
```

Comment:
Can't you just use the Function api for static methods?

Reply:
Not really, because they would not be placed correctly in Python since static methods in C++ should become methods on the class in Python. But you could if you don't care about that.